### PR TITLE
Report builder - Iterating using indexes of array and keys of an object

### PR DESCRIPTION
**Pull request by Arik-Cappsool**
Before this edit, for this report request:

report.getReport('v201705', {
reportName: 'te',
reportType: 'KEYWORDS_PERFORMANCE_REPORT',
fields: ["Date",'Criteria', 'Cost', 'CampaignName',"Impressions" , "Clicks"],
startDate: new Date("07/07/2017"),
endDate: new Date(),
additionalHeaders: {
skipReportHeader: true,
skipReportSummary: true
},
filters: [{field: 'Cost', operator: 'GREATER_THAN', values: [0]}],
format: 'CSV' //defaults to CSV
}, (error, report) => {
if(error) return callback(error);
return callback(null,report);
});
}

The generated XML was:

{ __rdxml: 'DateCriteriaCostCampaignNameImpressionsClicksfunction (callback) {\n var i = 0;\n while (i < this.length) {\n if (callback(this[i], i)) {\n this.splice(i, 1);\n }\n else {\n ++i;\n }\n }\n}function (obj) {\n var i = this.length;\n while (i--) {\n if (this[i] === obj) {\n return true;\n }\n }\n return false;\n}CostGREATER_THAN0function (callback) {\n var i = 0;\n while (i < this.length) {\n if (callback(this[i], i)) {\n this.splice(i, 1);\n }\n else {\n ++i;\n }\n }\n}function (obj) {\n var i = this.length;\n while (i--) {\n if (this[i] === obj) {\n return true;\n }\n }\n return false;\n}function (callback) {\n var i = 0;\n while (i < this.length) {\n if (callback(this[i], i)) {\n this.splice(i, 1);\n }\n else {\n ++i;\n }\n }\n}function (obj) {\n var i = this.length;\n while (i--) {\n if (this[i] === obj) {\n return true;\n }\n }\n return false;\n}function (callback) {\n var i = 0;\n while (i < this.length) {\n if (callback(this[i], i)) {\n this.splice(i, 1);\n }\n else {\n ++i;\n }\n }\n}function (obj) {\n var i = this.length;\n while (i--) {\n if (this[i] === obj) {\n return true;\n }\n }\n return false;\n}2017070720170724teKEYWORDS_PERFORMANCE_REPORTCUSTOM_DATECSV' }

### CODE REVIEW

**Comment by ChrisAlvares** [comment]
Hey @ArikCapp can you write a test to prove this out. I was unable to re-create on my end.

**Comment by ArikCapp** [comment]
Hey!
So I figured out what happens, We have legacy code that adds 2 functions to the Array prototype (Which  is maybe not common but works for us). Now when iterating over an array with the "var index in array"
notating, we get these functions.
So to reproduce it do something like this:

Array.prototype.contains = function(obj) {
  var i = this.length;
  while (i--) {
    if (this[i] === obj) {
      return true;
    }
  }
  return false;
}

and now try to create a reports. 
Iterating with indexes like in my PR solves it.

Thanks for your helpful package

**Comment by CMarker** [comment]
+1 to this. Had to fork and put this change in to be able to generate reports after having a prototype function set on Array (like the previous poster).

**Comment by sagitshemesh** [comment]
Hi, I also encountered this issue. It would help if you could merge this solution into the main project so we can keep working with it instead of working with the forked copies, and get updated any time you release new versions.
It's a simple fix so I don't think it should be complicated (adding hasOwnProperty() can also help).
Thanks!

File: adwords/report-builder.js [hunk 1]
```diff
@@ -48,7 +48,7 @@ class AdwordsReportBuilder {
      * @param fields {array}
      */
     buildFields(selector, fields) {
-        for (var index in fields) {
+        for (var index = 0; index < fields.length; index++) {
```
Comment (ChrisAlvares) [inline adwords/report-builder.js]: There are times where fields / filters / filter values are null. using .length without checking if they exist will end up causing code to crash. Can you switch to maybe a lodash `each` or use `(fields || []).forEach`.

**Comment by ChrisAlvares** [comment]
Closing as this will be added in the latest version

**Pull request closed without merging**